File: apps/ios/monocr-ios/PageNormalizer.swift
import Foundation

/**
 Put a page into the polarity the model was trained on, once, before anything
 reads it.

 The bug this fixes: the segmenter has no polarity handling at all, and inversion
 used to happen per line inside `ImagePreprocessor`, AFTER segmentation had
 already run on the un-inverted grey. On a dark-mode screenshot or an inverted
 scan the segmenter therefore measured the BACKGROUND as ink, so the lines it
 returned were the gaps. Inverting later cannot undo that.

 Ported from mon_OCR `src/monocr/utils.py` (`to_normalized_grayscale` and
 `_level_background`). It is not idempotent — running it per line after running
 it per page would level an already-levelled crop — so the per-line inversion was
 removed in the same change that added this.
 */
nonisolated enum PageNormalizer {

    /// Background luminance below this reads as a dark background to invert.
    private static let darkBackgroundMedian = 128.0

    /// Corner patch size, as a fraction of each page dimension, with a 3px floor
    /// so a tiny image still samples something.
    private static let cornerFraction = 10
    private static let cornerFloor = 3

    /// Dark text on a light, level background, whatever the input looked like.
    static func normalize(_ page: GreyImage) -> GreyImage {
        guard page.width > 0, page.height > 0 else { return page }

        let upright = backgroundIsDark(page) ? inverted(page) : page
        return levelBackground(upright)
    }

    /**
     Is the page light text on a dark background?

     Sampled from the four corner patches rather than from a global mean: page
     corners are almost always background, so their median survives a dense,
     text-heavy page where the mean would be dragged down by ink.
     */
    static func backgroundIsDark(_ page: GreyImage) -> Bool {
        guard page.width > 0, page.height > 0 else { return false }

        let patchH = min(max(cornerFloor, page.height / cornerFraction), page.height)
        let patchW = min(max(cornerFloor, page.width / cornerFraction), page.width)

        // Counting sort rather than a real sort: the four patches together can be
        // 4% of a 12MP photo, and only two order statistics are needed.
        var histogram = [Int](repeating: 0, count: 256)
        var count = 0
        let rowBands = [0..<patchH, (page.height - patchH)..<page.height]
        let colBands = [0..<patchW, (page.width - patchW)..<page.width]
        for rows in rowBands {
            for cols in colBands {
                for y in rows {
                    for x in cols {
                        histogram[Int(page.pixel(x: x, y: y))] += 1
                        count += 1
                    }
                }
            }
        }
        guard count > 0 else { return false }

        // numpy's median averages the two middle values of an even sample, and
        // this sample is always even (four patches of equal size).
        let lowerRank = (count - 1) / 2
        let upperRank = count / 2
        var seen = 0
        // -1, not 0, as the "not found yet" sentinel. 0 is a legal luminance — a
        // genuinely black corner — so using it for both meanings meant the sentinel
        // never cleared on such a page and `lower` was reassigned once more, coming
        // out 1 instead of 0. Measured on the web port, which was a line-for-line
        // copy of this: a page half black and half white reported median 128 against
        // a true 127.5, landing on the wrong side of `darkBackgroundMedian` — so the
        // inverted scan this function exists to catch was read as a light page.
        var lower = -1
        var upper = -1
        for value in 0..<256 {
            seen += histogram[value]
            if lower == -1 && seen > lowerRank { lower = value }
            if seen > upperRank {
                upper = value
                break
            }
        }
        let median = (Double(lower) + Double(upper)) / 2.0
        return median < darkBackgroundMedian
    }

    static func inverted(_ page: GreyImage) -> GreyImage {
        GreyImage(pixels: page.pixels.map { 255 - $0 }, width: page.width, height: page.height)
    }

    /**
     Flatten grey boxes, sepia paper and coloured panels to near-white while
     leaving ink dark.

     Dilation is a max filter, and background pixels are the bright ones, so
     dilating with a kernel far wider than a glyph propagates the local paper
     tone across the ink and yields a smooth background estimate. Dividing by
     that estimate maps any background shade to 255. Already-white pages come
     back unchanged.
     */
    static func levelBackground(_ page: GreyImage) -> GreyImage {
        guard page.width > 0, page.height > 0 else { return page }

        // Downsampled first so the kernel covers proportionally more of the page
        // and the max filter runs on 1/16 of the pixels.
        let smallH = max(1, page.height / 4)
        let smallW = max(1, page.width / 4)
        let small = downsampleArea(page, width: smallW, height: smallH)

        // Kernel covers ~10% of the small image height, forced odd: the upstream
        // balance between smoothing box edges and keeping global brightness
        // gradients.
        let kernel = max(7, (smallH / 4) | 1)
        let background = dilate(small, kernel: kernel)
        let estimate = upsampleBilinear(background, width: page.width, height: page.height)

        var levelled = [UInt8](repeating: 0, count: page.width * page.height)
        for i in 0..<levelled.count {
            // A near-black estimate would blow the quotient up; polarity is
            // already corrected by the time we get here, so clamping at 1 only
            // guards the arithmetic.
            let bg = max(estimate[i], 1.0)
            let value = Float(page.pixels[i]) / bg * 255.0
            levelled[i] = UInt8(max(0.0, min(255.0, value)))
        }
        return GreyImage(pixels: levelled, width: page.width, height: page.height)
    }

    // MARK: - Resampling and morphology

    /// Area-average downsample, matching OpenCV's INTER_AREA: each destination
    /// pixel is the mean of the source rectangle it covers, fractional edges
    /// weighted by overlap.
    static func downsampleArea(_ src: GreyImage, width dstW: Int, height dstH: Int) -> GreyImage {
        guard dstW > 0, dstH > 0, src.width > 0, src.height > 0 else {
            return GreyImage(pixels: [], width: 0, height: 0)
        }

        let scaleX = Double(src.width) / Double(dstW)
        let scaleY = Double(src.height) / Double(dstH)
        var out = [UInt8](repeating: 0, count: dstW * dstH)

        for dy in 0..<dstH {
            let y0 = Double(dy) * scaleY
            let y1 = min(Double(dy + 1) * scaleY, Double(src.height))
            let iy0 = min(Int(y0), src.height - 1)
            let iy1 = min(Int(y1.rounded(.up)), src.height)
            for dx in 0..<dstW {
                let x0 = Double(dx) * scaleX
                let x1 = min(Double(dx + 1) * scaleX, Double(src.width))
                let ix0 = min(Int(x0), src.width - 1)
                let ix1 = min(Int(x1.rounded(.up)), src.width)

                var sum = 0.0
                var weight = 0.0
                for y in iy0..<max(iy1, iy0 + 1) {
                    let wy = min(Double(y + 1), y1) - max(Double(y), y0)
                    if wy <= 0 { continue }
                    for x in ix0..<max(ix1, ix0 + 1) {
                        let wx = min(Double(x + 1), x1) - max(Double(x), x0)
                        if wx <= 0 { continue }
                        sum += wx * wy * Double(src.pixel(x: x, y: y))
                        weight += wx * wy
                    }
                }
                let mean = weight > 0 ? sum / weight : Double(src.pixel(x: ix0, y: iy0))
                out[dy * dstW + dx] = UInt8(max(0.0, min(255.0, mean.rounded())))
            }
        }
        return GreyImage(pixels: out, width: dstW, height: dstH)
    }

    /**
     Dilate with a DISK, matching `cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (k, k))`.

     **This was a square until 2026-08-28, and that was wrong.** The reference
     (`mon_OCR/src/monocr/utils.py`, `_level_background`) asks cv2 for `MORPH_ELLIPSE`,
     and Android's `PageNormalizer.dilateDisk` has always matched it. A square contains
     the inscribed disk, so the square propagated more background over the ink, the
     estimate came back brighter, and dividing by it made every iOS page 0.13%-0.34%
     darker than the same page on Android. Measured over eight synthetic pages, seven
     disagreed; only the 1x1 case matched.

     The old test did not catch it because it compared this function against a naive
     max filter that was also a square: it pinned the sliding-window optimisation
     correctly and never asked what shape was being optimised.

     cv2's ellipse is not the geometric disk `dx^2 + dy^2 <= r^2` — that disagrees with
     it at every kernel size tested. It is the per-row half-width `round(sqrt(r^2 - dy^2))`,
     which agrees at all of them, so that is what this builds.

     Ported from Android's implementation, which is the one already verified against
     cv2, rather than rewritten: two independent derivations of the same shape is how
     the ports drifted apart in the first place.

     **It costs more than the square did, and the old comment here was right to say
     so.** A square max filter is separable, so two monotonic-deque passes cost the
     same whatever the radius; a disk needs one horizontal growth pass per radius
     plus one vertical fold per kernel row. Measured on this machine with `swiftc -O`
     at the size a 12 MP photo actually produces — 3024x4032 downsamples to 756x1008,
     and `max(7, (1008 / 4) | 1)` gives a kernel of 253:

         square, separable      10.5 ms
         disk                  125.8 ms

     That is the price of computing what the reference computes, and Android has
     always paid it — its own comment budgets "a few hundred milliseconds" for this
     function on a 300 DPI page. The old comment used this cost to justify the square;
     the cost was real and the conclusion was not, because it traded a correctness
     property nothing was measuring for milliseconds nothing was waiting on.

     The 125.8 ms is with `UInt8` buffers rather than `Int` (a third of the memory
     traffic: three 762k arrays at 1 byte instead of 8), `swap` rather than array
     assignment, and clipped `y` ranges instead of a per-row bounds test. The
     straightforward form measured 188.5 ms and produces bit-identical output.
     */
    static func dilate(_ src: GreyImage, kernel: Int) -> GreyImage {
        let r = max(0, kernel / 2)
        guard r > 0, src.width > 0, src.height > 0 else { return src }

        let w = src.width
        let h = src.height

        // Half-width of the disk on each row of the structuring element.
        var halfWidth = [Int](repeating: 0, count: 2 * r + 1)
        for i in 0...(2 * r) {
            let dy = Double(i - r)
            halfWidth[i] = Int((Double(r * r) - dy * dy).squareRoot().rounded())
        }

        // `current` holds src dilated horizontally by d, grown one step per pass.
        // 0 is the identity for a maximum over 0...255 data, so no sentinel is needed
        // and these can stay UInt8.
        var current = src.pixels
        var next = [UInt8](repeating: 0, count: w * h)
        var out = [UInt8](repeating: 0, count: w * h)

        for d in 0...r {
            if d >= 1 && w > 1 {
                current.withUnsafeBufferPointer { cur in
                    next.withUnsafeMutableBufferPointer { nxt in
                        for y in 0..<h {
                            let base = y * w
                            // Half-width d covers [x-d, x+d]. For d == 1 that needs
                            // the centre too; from d == 2 on, the two half-width
                            // d-1 windows either side of x already cover it.
                            var first: UInt8 = d == 1 ? cur[base] : 0
                            if cur[base + 1] > first { first = cur[base + 1] }
                            nxt[base] = first

                            if w > 2 {
                                for x in 1..<(w - 1) {
                                    var m: UInt8 = d == 1 ? cur[base + x] : 0
                                    let left = cur[base + x - 1]
                                    let right = cur[base + x + 1]
                                    if left > m { m = left }
                                    if right > m { m = right }
                                    nxt[base + x] = m
                                }
                            }

                            var last: UInt8 = d == 1 ? cur[base + w - 1] : 0
                            if cur[base + w - 2] > last { last = cur[base + w - 2] }
                            nxt[base + w - 1] = last
                        }
                    }
                }
                // `next` is fully overwritten on every pass, so swapping is safe and
                // avoids the copy an assignment would make.
                swap(&current, &next)
            }

            for i in 0...(2 * r) where halfWidth[i] == d {
                let dy = i - r
                let yLo = max(0, -dy)
                let yHi = min(h, h - dy)
                if yLo >= yHi { continue }
                current.withUnsafeBufferPointer { cur in
                    out.withUnsafeMutableBufferPointer { o in
                        for y in yLo..<yHi {
                            let dstRow = y * w
                            let srcRow = (y + dy) * w
                            for x in 0..<w {
                                let v = cur[srcRow + x]
                                if v > o[dstRow + x] { o[dstRow + x] = v }
                            }
                        }
                    }
                }
            }
        }

        return GreyImage(pixels: out, width: w, height: h)
    }

    static func upsampleBilinear(_ src: GreyImage, width: Int, height: Int) -> [Float] {
        var out = [Float](repeating: 0, count: width * height)
        guard src.width > 0, src.height > 0, width > 0, height > 0 else { return out }

        let scaleX = Double(src.width) / Double(width)
        let scaleY = Double(src.height) / Double(height)

        for y in 0..<height {
            let sy = max(0.0, (Double(y) + 0.5) * scaleY - 0.5)
            let y0 = min(Int(sy), src.height - 1)
            let y1 = min(y0 + 1, src.height - 1)
            let fy = Float(sy - Double(y0))
            for x in 0..<width {
                let sx = max(0.0, (Double(x) + 0.5) * scaleX - 0.5)
                let x0 = min(Int(sx), src.width - 1)
                let x1 = min(x0 + 1, src.width - 1)
                let fx = Float(sx - Double(x0))

                let top = Float(src.pixel(x: x0, y: y0)) * (1 - fx) + Float(src.pixel(x: x1, y: y0)) * fx
                let bottom = Float(src.pixel(x: x0, y: y1)) * (1 - fx) + Float(src.pixel(x: x1, y: y1)) * fx
                out[y * width + x] = top * (1 - fy) + bottom * fy
            }
        }
        return out
    }
}

File: apps/android/app/src/main/java/dev/janakhpon/monocr/engine/PageNormalizer.kt
package dev.janakhpon.monocr.engine

import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.sqrt

/**
 * Puts a page into the polarity and background the model was trained on: dark ink
 * on a white ground.
 *
 * This runs once, on the whole page, before segmentation. It used to run per line
 * inside the preprocessor, after segmentation had already measured ink on the
 * un-inverted grey — so on a dark-mode screenshot or an inverted scan the segmenter
 * found the background and called it text, and no amount of correct per-line
 * inversion afterwards could recover the lines it had already missed.
 *
 * It is not idempotent: the background levelling divides by an estimate of the
 * background, so running it twice divides twice. There is exactly one call site,
 * in [OcrRepository.performOcr], and the per-line inversion it replaced is gone.
 *
 * Ported from mon_OCR `utils.to_normalized_grayscale` and `utils._level_background`.
 * That port is faithful to the algorithm, not bit-exact to OpenCV: the area resize,
 * the elliptical dilation and the bilinear upsample are written out here in
 * floating point, where OpenCV uses fixed-point kernels. Expect agreement to about
 * a grey level, not to the bit.
 */
object PageNormalizer {

    // `backgroundIsDark`, `invert`, `levelBackground`, `resizeArea` and `dilateDisk`
    // are deliberately not private. Only `normalize` was reachable until 2026-08-28,
    // and the consequence was that the most numerically delicate file in `engine/`
    // had zero tests while the iOS port of the same code had eight — iOS having
    // widened exactly these for exactly this reason.
    //
    // What that cost: iOS's dilation used a square structuring element where this
    // one uses a disk, so the two ports disagreed on 7 of 8 synthetic pages, and
    // nothing on either side could see it. `buildLinearTable` and `areaWeights` stay
    // private; they are reached through the two functions above.

    /**
     * The dilation kernel for a downsampled page of [smallHeight] rows.
     *
     * `max(7, (h / 4) | 1)`: about 10% of the small image's height, forced odd so the
     * structuring element has a centre, with a floor that keeps it wider than a stroke
     * on small inputs. Named rather than inlined so `dilate-cases.json` can pin it
     * against the reference's own answers — a mutation dropping the floor from 7 to 3
     * survived the entire suite on 2026-08-28 because nothing asserted it.
     */
    fun kernelForSmallHeight(smallHeight: Int): Int = maxOf(7, (smallHeight / 4) or 1)

    /**
     * The corner-patch side for an image [side] pixels across that axis.
     *
     * `max(3, side / 10)`, matching `to_normalized_grayscale`. The floor of 3 is what
     * keeps a patch on a tiny crop from collapsing to a single pixel, and it was
     * likewise unpinned until `dilate-cases.json` carried the reference's answers.
     * The caller clamps to the axis length; on a small image the four patches overlap
     * and the shared pixels are counted twice, which is what the reference does.
     */
    fun cornerPatch(side: Int): Int = maxOf(3, side / 10)

    /** Corner patches below this median read as a dark background, so the page is inverted. */
    private const val DARK_BACKGROUND_MEDIAN = 128.0

    /**
     * Normalise [page] in place and return it. In place because both steps write each
     * output pixel from the same input pixel, and a full-size copy per step is 33 MB
     * on a 300 DPI A4 page.
     */
    fun normalize(page: GreyImage): GreyImage {
        if (page.width == 0 || page.height == 0) return page
        if (backgroundIsDark(page)) invert(page)
        levelBackground(page)
        return page
    }

    /**
     * Sample the four 10%-corner patches and take their median.
     *
     * Document corners are almost always background, so their median tracks
     * background luminance even on a dense, text-heavy page where a global mean
     * would be dragged down by ink.
     */
    fun backgroundIsDark(page: GreyImage): Boolean {
        val h = page.height
        val w = page.width
        val ch = minOf(h, cornerPatch(h))
        val cw = minOf(w, cornerPatch(w))

        // A 256-bin histogram gives the exact median in one pass; sorting the
        // concatenated patches would be a third of a million comparisons on a
        // 300 DPI page for the same answer.
        val histogram = IntArray(256)
        var total = 0L
        for (yRange in listOf(0 until ch, (h - ch) until h)) {
            for (xRange in listOf(0 until cw, (w - cw) until w)) {
                for (y in yRange) {
                    val row = y * w
                    for (x in xRange) {
                        histogram[page.pixels[row + x]]++
                        total++
                    }
                }
            }
        }
        if (total == 0L) return false

        // Four patches of equal size, so the count is even and the median is the
        // mean of the two middle values, as numpy.median gives it.
        val lowerRank = total / 2 - 1
        val upperRank = total / 2
        var seen = 0L
        var lower = -1
        var upper = -1
        for (value in 0..255) {
            seen += histogram[value]
            if (lower < 0 && seen > lowerRank) lower = value
            if (seen > upperRank) {
                upper = value
                break
            }
        }
        return (lower + upper) / 2.0 < DARK_BACKGROUND_MEDIAN
    }

    fun invert(page: GreyImage) {
        val pixels = page.pixels
        for (i in pixels.indices) pixels[i] = 255 - pixels[i]
    }

    /**
     * Flatten grey boxes, off-white paper and coloured panels to white while leaving
     * ink dark.
     *
     * Because background is bright and ink is dark, a max filter (dilation) with a
     * kernel wider than a glyph propagates background over the ink and leaves a
     * smooth estimate of the local background tone. Dividing by that estimate maps
     * any background shade to white. Already-white pages come out unchanged.
     *
     * The estimate is built at quarter resolution, which is where the cost sits: on
     * a 300 DPI A4 page this is a few hundred milliseconds, against tens of
     * milliseconds of model time per line. It is not free, and the upstream
     * docstring's "0.5 ms" is an OpenCV number, not this one.
     */
    fun levelBackground(page: GreyImage) {
        val h = page.height
        val w = page.width
        val smallH = maxOf(1, h / 4)
        val smallW = maxOf(1, w / 4)

        val small = resizeArea(page, smallW, smallH)
        // Kernel covers about 10% of the small image height, forced odd so the
        // structuring element has a centre. The floor of 7 keeps it wider than a
        // stroke on small inputs.
        val kernel = kernelForSmallHeight(smallH)
        val backgroundSmall = dilateDisk(small, smallW, smallH, kernel)

        // The bilinear upsample and the division are fused: a full-resolution copy of
        // the background estimate would be another w*h floats, 33 MB on a 300 DPI
        // page, and each output row only needs two rows of the estimate.
        val sxIndex = IntArray(w)
        val sxFrac = FloatArray(w)
        buildLinearTable(smallW, w, sxIndex, sxFrac)
        val syIndex = IntArray(h)
        val syFrac = FloatArray(h)
        buildLinearTable(smallH, h, syIndex, syFrac)

        // Written in place: each output pixel reads only its own input pixel and the
        // already-finished background estimate, so nothing downstream in this loop
        // needs the original value.
        val out = page.pixels
        for (y in 0 until h) {
            val y0 = syIndex[y]
            val wy = syFrac[y]
            val y1 = if (wy == 0f) y0 else y0 + 1
            val row0 = y0 * smallW
            val row1 = y1 * smallW
            val dstRow = y * w
            for (x in 0 until w) {
                val x0 = sxIndex[x]
                val wx = sxFrac[x]
                val x1 = if (wx == 0f) x0 else x0 + 1
                val top = backgroundSmall[row0 + x0] * (1f - wx) + backgroundSmall[row0 + x1] * wx
                val bottom = backgroundSmall[row1 + x0] * (1f - wx) + backgroundSmall[row1 + x1] * wx
                // Guard the division rather than trusting the estimate to stay bright:
                // an all-black input survives polarity normalisation as all-black.
                val background = maxOf(top * (1f - wy) + bottom * wy, 1f)
                val scaled = (out[dstRow + x] / background) * 255f
                out[dstRow + x] = scaled.coerceIn(0f, 255f).toInt()
            }
        }
    }

    /**
     * OpenCV's `INTER_LINEAR` sample positions: pixel centres at half-integers, the
     * source index clamped to the last column and its weight dropped there so the
     * edge replicates instead of reading past the buffer.
     */
    private fun buildLinearTable(srcLen: Int, dstLen: Int, index: IntArray, frac: FloatArray) {
        val scale = srcLen.toDouble() / dstLen
        for (d in 0 until dstLen) {
            val position = (d + 0.5) * scale - 0.5
            var s = floor(position).toInt()
            var a = (position - s).toFloat()
            if (s < 0) {
                s = 0
                a = 0f
            }
            if (s >= srcLen - 1) {
                s = srcLen - 1
                a = 0f
            }
            index[d] = s
            frac[d] = a
        }
    }

    /**
     * Area-averaged downscale, the separable equivalent of `cv2.INTER_AREA`. Each
     * output pixel is the mean of the source pixels its footprint covers, weighted
     * by how much of each it covers. Rounded back to 0..255 because `cv2.resize` on
     * a uint8 image returns uint8, and the dilation below then sees the same values
     * the reference implementation dilates.
     */
    fun resizeArea(page: GreyImage, dstW: Int, dstH: Int): IntArray {
        val srcW = page.width
        val srcH = page.height
        val xWeights = areaWeights(srcW, dstW)
        val yWeights = areaWeights(srcH, dstH)

        val horizontal = FloatArray(dstW * srcH)
        for (y in 0 until srcH) {
            val srcRow = y * srcW
            val dstRow = y * dstW
            for (dx in 0 until dstW) {
                val weights = xWeights.weights[dx]
                val first = xWeights.first[dx]
                var acc = 0f
                for (i in weights.indices) acc += weights[i] * page.pixels[srcRow + first + i]
                horizontal[dstRow + dx] = acc
            }
        }

        val out = IntArray(dstW * dstH)
        for (dy in 0 until dstH) {
            val weights = yWeights.weights[dy]
            val first = yWeights.first[dy]
            val dstRow = dy * dstW
            for (x in 0 until dstW) {
                var acc = 0f
                for (i in weights.indices) acc += weights[i] * horizontal[(first + i) * dstW + x]
                out[dstRow + x] = Math.round(acc).coerceIn(0, 255)
            }
        }
        return out
    }

    private class AreaWeights(val first: IntArray, val weights: Array<FloatArray>)

    private fun areaWeights(srcLen: Int, dstLen: Int): AreaWeights {
        val scale = srcLen.toDouble() / dstLen
        val first = IntArray(dstLen)
        val weights = Array(dstLen) { FloatArray(0) }
        for (d in 0 until dstLen) {
            val start = d * scale
            val end = minOf((d + 1) * scale, srcLen.toDouble())
            val i0 = start.toInt()
            val i1 = maxOf(i0 + 1, minOf(ceil(end).toInt(), srcLen))
            val row = FloatArray(i1 - i0)
            var total = 0.0
            for (i in i0 until i1) {
                val overlap = maxOf(0.0, minOf(end, (i + 1).toDouble()) - maxOf(start, i.toDouble()))
                row[i - i0] = overlap.toFloat()
                total += overlap
            }
            if (total > 0.0) {
                for (i in row.indices) row[i] = (row[i] / total).toFloat()
            } else {
                row[0] = 1f
            }
            first[d] = i0
            weights[d] = row
        }
        return AreaWeights(first, weights)
    }

    /**
     * Dilate with a disk of diameter [k], matching
     * `cv2.getStructuringElement(MORPH_ELLIPSE, (k, k))` then `cv2.dilate`.
     *
     * The disk is a union of horizontal segments, one per row offset, so the
     * dilation is the maximum over those segments. Growing the horizontal maximum
     * one pixel per step and folding in each row offset as its width is reached
     * makes this O((r + k) * w * h). The direct k-by-k scan is O(k^2) per pixel,
     * which on a 300 DPI page means a kernel around 200 wide and tens of billions of
     * comparisons — not slow, unusable.
     *
     * Out-of-image parts of the kernel are skipped, which is what OpenCV's default
     * dilation border does.
     */
    fun dilateDisk(src: IntArray, w: Int, h: Int, k: Int): IntArray {
        val r = k / 2
        val invR2 = if (r > 0) 1.0 / (r.toDouble() * r) else 0.0
        val halfWidth = IntArray(2 * r + 1)
        for (i in 0..2 * r) {
            val dy = (i - r).toDouble()
            halfWidth[i] = Math.round(r * sqrt((r.toDouble() * r - dy * dy) * invR2)).toInt()
        }

        var current = src.copyOf()
        val next = IntArray(w * h)
        // 0 is the identity for a maximum over 0..255 data, so no sentinel is needed.
        val out = IntArray(w * h)

        for (d in 0..r) {
            if (d >= 1) {
                for (y in 0 until h) {
                    val base = y * w
                    for (x in 0 until w) {
                        // Half-width d covers [x-d, x+d]. For d == 1 that needs the
                        // centre too; from d == 2 on, the two half-width d-1 windows
                        // either side of x already overlap and cover it.
                        var m = if (d == 1) current[base + x] else Int.MIN_VALUE
                        if (x > 0) m = maxOf(m, current[base + x - 1])
                        if (x + 1 < w) m = maxOf(m, current[base + x + 1])
                        next[base + x] = if (m == Int.MIN_VALUE) current[base + x] else m
                    }
                }
                System.arraycopy(next, 0, current, 0, current.size)
            }
            for (i in 0..2 * r) {
                if (halfWidth[i] != d) continue
                val dy = i - r
                for (y in 0 until h) {
                    val sy = y + dy
                    if (sy < 0 || sy >= h) continue
                    val dstRow = y * w
                    val srcRow = sy * w
                    for (x in 0 until w) {
                        val v = current[srcRow + x]
                        if (v > out[dstRow + x]) out[dstRow + x] = v
                    }
                }
            }
        }
        return out
    }
}
